Remotion/Web/Core/Context/HttpContextStorageProviderNetCore.cs

```diff
@@ -0,0 +1,182 @@
+      {
+        if (TryGetCachedHashtable(context, out var hashtable))
+        {
+          return hashtable;
+        }
+
+        return SetHashtable(context, new Hashtable());
+      }
+
+      public Hashtable? RemoveHashtable (HttpContext context)
+      {
+        if (TryGetCachedHashtable(context, out var cachedHashtable))
+        {
+          context.Items.Remove(c_dataDictionaryKey);
+          _dataTable.Value!.SetTarget(null!);
+        }
+
+        return cachedHashtable;
+      }
+
+      public Hashtable SetHashtable (HttpContext context, Hashtable hashtable)
+      {
+        var wrappedHashtable = new WrappedHashtable(context, hashtable);
+        context.Items[c_dataDictionaryKey] = wrappedHashtable;
+        _dataTable.Value!.SetTarget(wrappedHashtable);
+        return hashtable;
+      }
+    }
+
+    // Ignore the obsolete warnings for AsyncLocalStorageProvider
+#pragma warning disable 618
+    private readonly AsyncLocalStorageProvider _fallbackProvider = new();
+#pragma warning restore 618
+
+    private readonly HashTableAccessor _hashTableAccessor = new(true);
+
+    public SafeContextBoundary OpenSafeContextBoundary ()
+    {
+      var httpContext = HttpContext.Current;
+
+      if (httpContext == null)
+        return _fallbackProvider.OpenSafeContextBoundary();
+
+      var hashtable = _hashTableAccessor.RemoveHashtable(httpContext);
+
+      return new SafeContextBoundary(this, hashtable);
+    }
+
+    public object? GetData (string key)
+    {
+      ArgumentUtility.CheckNotNull("key", key);
+
+      var httpContext = HttpContext.Current;
+
+      if (httpContext == null)
+        return _fallbackProvider.GetData(key);
+
+      if (_hashTableAccessor.TryGetCachedHashtable(httpContext, out var hashtable))
+        return hashtable[key];
+
+      return null;
+    }
+
+    public void SetData (string key, object? value)
+    {
+      ArgumentUtility.CheckNotNull("key", key);
+
+      var httpContext = HttpContext.Current;
+
+      if (httpContext == null)
+      {
+        _fallbackProvider.SetData(key, value);
+        return;
+      }
+
+      var hashtable = _hashTableAccessor.GetOrCreateHashtable(httpContext);
+      hashtable[key] = value;
+    }
+
+    public void FreeData (string key)
+    {
+      ArgumentUtility.CheckNotNull("key", key);
+
+      var httpContext = HttpContext.Current;
+
+      if (httpContext == null)
+      {
+        _fallbackProvider.FreeData(key);
+        return;
+      }
+
+      if (_hashTableAccessor.TryGetCachedHashtable(httpContext, out var cachedHashtable))
+        cachedHashtable.Remove(key);
+    }
+
+    void ISafeContextBoundaryStrategy.RestorePreviousSafeContext (object? state)
+    {
+      var httpContext = HttpContext.Current;
+
+      if (state is Hashtable hashtable)
+        _hashTableAccessor.SetHashtable(httpContext, hashtable);
+      else if (_hashTableAccessor.TryGetCachedHashtable(httpContext, out _))
+        _hashTableAccessor.RemoveHashtable(httpContext);
```

Comment:
`HashtableAccessor.RemoveHashtable` already does the check for us we can simplify this to an `else` if I am not mistaken.